### PR TITLE
feat: [CDS-72825]: ssh/winrm skip copy/download artifact for githubPackages for container and docker

### PR DESCRIPTION
**Pull request by vitsafronovici**
## Describe your changes
For ssh/winrm deployments an error hint is shown when try to copy/download artifact for githubPackages for container packageType

image

## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49672)

### CODE REVIEW

**Comment by vitsafronovici** [comment]
trigger uts

**Comment by vitsafronovici** [comment]
trigger messagecheck

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java [hunk 1]
```diff
@@ -133,6 +124,39 @@ private CommandExecutionStatus downloadArtifact(CommandTaskParameters commandTas
     }
   }
 
+  private void verifyIfDownloadIsSupported(SshWinRmArtifactDelegateConfig artifactDelegateConfig) {
```
Comment (acasian-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java:127]: for download & copy (have a single method) i believe we need to extract the code and move upper in `CommandHandler` and leverage in both command handlers.

Comment (acasian-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java:127]: nit: method name could be `validateArtifactType`

**Review by acasian-harness — CHANGES_REQUESTED** [review]

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49672)* status: Some of the checks are too old. 5 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java` line 127 at r1](https://reviewable.io/reviews/harness/harness-core/49672#gh-1243502096:-NZ333nl55WsNnOZYIwl:bhhhjv3) ([raw file](https://github.com/harness/harness-core/blob/7439896df8871a9bff78de8a63de21729eb71ffd/930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java#L127)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

nit: method name could be `validateArtifactType`
</blockquote></details>

Agree, we can leverage the same method and move to the upper handler if the messages are the same. 
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java` line 148 at r1](https://reviewable.io/reviews/harness/harness-core/49672#-NZ31qTi0Nvn5X1AQug6:-NZ31qTi0Nvn5X1AQug7:b-uixuzj) ([raw file](https://github.com/harness/harness-core/blob/7439896df8871a9bff78de8a63de21729eb71ffd/930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java#L148)):*
> ```Java
>       GithubPackagesArtifactDelegateConfig githubPackagesArtifactDelegateConfig =
>           (GithubPackagesArtifactDelegateConfig) artifactDelegateConfig;
>       if (!GithubPackageUtils.MAVEN_PACKAGE_TYPE.equals(githubPackagesArtifactDelegateConfig.getPackageType())) {
> ```

It's better to specify the types that are not supported if it's doable
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/ssh/exception/SshExceptionConstants.java` line 139 at r1](https://reviewable.io/reviews/harness/harness-core/49672#-NZ32YmWByXBv0h2MhNr:-NZ32YmWByXBv0h2MhNs:bbjtwmb) ([raw file](https://github.com/harness/harness-core/blob/7439896df8871a9bff78de8a63de21729eb71ffd/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/ssh/exception/SshExceptionConstants.java#L139)):*
> ```Java
>       "Failed to download GithubPackage Artifact with id: %s";
> 
>   public final String DOWNLOAD_ARTIFACT_NOT_SUPPORTED_FOR_GITHUB_PACKAGE_ARTIFACT_HINT =
> ```

Minor - update the constant name to contain and copy. It's better to have separate messages for downloading and copying. 
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/ssh/exception/SshExceptionConstants.java` line 141 at r1](https://reviewable.io/reviews/harness/harness-core/49672#-NZ32owu5SA-TYJQA9Yr:-NZ32owu5SA-TYJQA9Ys:b-64v23t) ([raw file](https://github.com/harness/harness-core/blob/7439896df8871a9bff78de8a63de21729eb71ffd/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/ssh/exception/SshExceptionConstants.java#L141)):*
> ```Java
>   public final String DOWNLOAD_ARTIFACT_NOT_SUPPORTED_FOR_GITHUB_PACKAGE_ARTIFACT_HINT =
>       "Please make sure there is no copy or download artifact command unit specified";
>   public final String DOWNLOAD_ARTIFACT_NOT_SUPPORTED_FOR_GITHUB_PACKAGE_ARTIFACT_EXPLANATION =
> ```

Minor - The same as above.
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/ssh/exception/SshExceptionConstants.java` line 143 at r1](https://reviewable.io/reviews/harness/harness-core/49672#-NZ32ryK8f0HAEBQzay1:-NZ32ryK8f0HAEBQzay2:b-64v23t) ([raw file](https://github.com/harness/harness-core/blob/7439896df8871a9bff78de8a63de21729eb71ffd/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/ssh/exception/SshExceptionConstants.java#L143)):*
> ```Java
>   public final String DOWNLOAD_ARTIFACT_NOT_SUPPORTED_FOR_GITHUB_PACKAGE_ARTIFACT_EXPLANATION =
>       "Copy and Download Artifact for package type `%s` is not supported for Github Package Repository artifacts defined in service";
>   public final String DOWNLOAD_ARTIFACT_NOT_SUPPORTED_FOR_GITHUB_PACKAGE_ARTIFACT_FAILED =
> ```

Minor - The same as above.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java (hunk 1 above)
Comment (vitsafronovici) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/handler/AbstractDownloadArtifactCommandHandler.java:127]: copy differs from download, if it's an instance of SkipCopyArtifactDelegateConfig then we return SUCCESS, but for download - we throw exception

**Comment by vitsafronovici** [comment]
trigger uts

**Review by ivan-83 — DISMISSED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49672)* status: 5 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Review by acasian-harness — APPROVED** [review]

**Comment by vitsafronovici** [comment]
trigger uts

**Comment by vitsafronovici** [comment]
trigger uts

**Comment by vitsafronovici** [comment]
trigger review

**Review by acasian-harness — APPROVED** [review]

**Comment by acasian-harness** [comment]
trigger review

**Pull request merged**